--- a/fragalyseqt.py	2024-09-06 23:23:58.000000000 +0200
+++ b/fragalyseqt.py	2025-02-25 20:00:28.548974871 +0100
@@ -7,13 +7,13 @@
 # or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 #
 # You should have received a copy of the GNU Affero General Public License along with FragalyseQt. If not, see <https://www.gnu.org/licenses/>.
-import boxes, localize, fillarray
+from FragalyseApp import boxes, localize, fillarray
 from os.path import expanduser, dirname
 #Using FileDialog and SpinBox from pyqtgraph to prevent some possible problems for macOS users and to allow more fine variable setting.
 from pyqtgraph import PlotWidget, FileDialog, SpinBox, ComboBox
 #Using widgets from pyqtgraph to make program independent from Qt for Python implementation.
 from pyqtgraph.Qt.QtWidgets import QCheckBox, QTableWidget, QTableWidgetItem
-from sizestandards import size_standards
+from FragalyseApp.sizestandards import size_standards
 ftype = "ABI fragment analysis files (*.fsa *.hid)"
 global show_channels, ifacemsg, do_BCD
 do_BCD = False
